documentation/testing_plans.md

```diff
@@ -0,0 +1,815 @@
+
+Given the following test which has three `expect_task` mocks:
+
+```ruby
+it 'runs the configure task' do
+  expect_task('configure')
+  expect_task('configure').with_targets(['servers'])
+  expect_task('configure').with_targets(['servers']).with_params('autoupdate' => true)
+
+  run_plan('configure', 'targets' => 'servers')
+end
+```
+
+`BoltSpec` matches the `run_task` invocation to the following mocks:
+
+```ruby
+expect_task('configure')
+expect_task('configure').with_targets(['servers'])
+```
+
+`BoltSpec` does not match the `run_task` invocation to the third mock because it
+uses the `with_params` modifier to add more specificity. Because `BoltSpec`
+found multiple matching mocks, it will use the last one defined in the test:
+`expect_task('configure').with_targets(['servers'])`. The test will then fail,
+as the first mock is not fulfilled, as the plan does not run the `configure`
+task a second time.
+
+It's important to remember that `BoltSpec` uses the last matching stub or mock
+and write your tests accordingly by creating your stubs and mocks in the correct
+order. Otherwise, your test might fail even though your plan's logic is sound.
+
+For example, the following plan runs the `configure` task twice. The first task
+run sets the `autoupdate` parameter, while the second task run does not.
+
+```puppet
+plan configure_twice (
+  TargetSpec $targets
+) {
+  run_task('configure', $targets, 'autoupdate' => true)
+  run_task('configure', $targets)
+}
+```
+
+The following test creates two mocks. The first mock asserts that the
+`configure` task is run with the parameter `autoupdate => true`, while the
+second mock just asserts that the `configure` task is run.
+
+```ruby
+it 'configures servers and databases' do
+  expect_task('configure').with_params('autoupdate' => true)
+  expect_task('configure')
+
+  run_plan('configure_twice', 'targets' => 'servers')
+end
+```
+
+When the test is run, `BoltSpec` matches the first `run_task` invocation to both
+mocks, but uses the last matching mock: `expect_task('configure')`. When the
+plan invokes `run_task` a second time, the invocation only matches one mock:
+`expect_task('configure')`. The test fails because the
+`expect_task('configure')` mock is only expected to be matched once and the test
+never used the first `expect_task('configure').with_params('autoupdate' =>
+true)` mock.
+
+Writing the test with the mocks in the opposite order results in the test
+passing:
+
+```ruby
+it 'configures servers and databases' do
+  expect_task('configure')
+  expect_task('configure').with_params('autoupdate' => true)
+
+  run_plan('configure_twice', 'targets' => 'servers')
+end
+```
+
+## Execution modes
+
+Plans often execute sub-plans with the `run_plan` function to build complex
+workflows. When testing these plans, it might be helpful to execute any
+sub-plans as well without needing to stub or mock the plan. To support this,
+`BoltSpec` offers two different execution modes:
+
+- `execute_any_plan`
```

Comment:
Ok, I *swear* I searched both this page and my checkout before commenting above :P Grep has failed me.